### PR TITLE
[C++ API Parity] Adam: updated step and class design 

### PR DESCRIPTION
**Pull request by anjali411**
**This PR is BC-breaking in the following way:**

In AdamOptions:
1. learning_rate is renamed to lr.
2. beta1 and beta2 are replaced by a tuple betas

**Test plan before 1.5 release:**

Test that in 1.5 we can load a C++ Adam optimizer that was serialized in 1.4, and their states are the same.

Test script:

optim_save_1_4.cpp
https://gist.github.com/yf225/b1c036c9b9c990ef20737a662bc927d7
optim_load_1_5.h
https://gist.github.com/yf225/84a47a40846b0eff6b87999986690fe0
optim_load_1_5.cpp
https://gist.github.com/yf225/44edfc6f4fe1a92be9232d402ece3714

### CODE REVIEW

File: torch/csrc/api/include/torch/optim/serialize.h [hunk 1]
```diff
@@ -219,8 +219,13 @@ void serialize(
 #define _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(T, name) \
 { \
   c10::IValue ivalue; \
-  archive.read(#name, ivalue); \
-  name(ivalue.to<T>()); \
+  bool exists = archive.try_read(#name, ivalue); \
+  if(exists) { \
+    name(ivalue.to<T>()); \
+  } else { \
+    /*undefined tensor*/ \
```
Comment (anjali411) [inline torch/csrc/api/include/torch/optim/serialize.h]: discussed offline: remove this and add default value({}) for arguments that may have undefined tensor

File: torch/csrc/api/src/optim/adam.cpp [hunk 2]
```diff
@@ -12,57 +12,171 @@
...
+          torch::equal(lhs.exp_avg(), rhs.exp_avg()) &&
+          torch::equal(lhs.exp_avg_sq(), rhs.exp_avg_sq()) &&
+          !(lhs.max_exp_avg_sq().defined() ^ rhs.max_exp_avg_sq().defined());
+}
+
+void AdamParamState::serialize(torch::serialize::OutputArchive& archive) const {
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(step);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(exp_avg);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(exp_avg_sq);
+  if(max_exp_avg_sq().defined()) {
```
Comment (yf225) [inline torch/csrc/api/src/optim/adam.cpp]: I think we should move the logic for skipping undefined tensor into `_TORCH_OPTIM_SERIALIZE_TORCH_ARG`, similar to what we do for `_TORCH_OPTIM_DESERIALIZE_TORCH_ARG`.

File: torch/csrc/api/src/optim/adam.cpp [hunk 3]
```diff
@@ -12,57 +12,171 @@
...
+      }
+      auto grad = p.grad().data();
+      TORCH_CHECK(!grad.is_sparse(), "Adam does not support sparse gradients"/*, please consider SparseAdam instead*/);
+      auto param_state = state_.find(c10::guts::to_string(p.unsafeGetTensorImpl()));
+      auto& options = static_cast<AdamOptions&>(group.options());
+      // State initialization
+      if(param_state == state_.end()) {
+        auto state = std::make_unique<AdamParamState>();
+        state->step(0);
+        state->exp_avg(torch::zeros_like(p.data(), MemoryFormat::Preserve));
```
Comment (yf225) [inline torch/csrc/api/src/optim/adam.cpp]: maybe add comment above: `// Exponential moving average of gradient values`

File: torch/csrc/api/src/optim/adam.cpp [hunk 4]
```diff
@@ -12,57 +12,171 @@
...
+        state->exp_avg_sq(torch::zeros_like(p.data(), MemoryFormat::Preserve));
```
Comment (yf225) [inline torch/csrc/api/src/optim/adam.cpp]: maybe add comment above: `// Exponential moving average of squared gradient values`

File: torch/csrc/api/src/optim/adam.cpp [hunk 5]
```diff
@@ -12,57 +12,171 @@
...
+        if(options.amsgrad()) {
+          state->max_exp_avg_sq(torch::zeros_like(p.data(), MemoryFormat::Preserve));
```
Comment (yf225) [inline torch/csrc/api/src/optim/adam.cpp]: maybe add comment above: `// Maintains max of all exp. moving avg. of sq. grad. values`

File: torch/csrc/api/src/optim/adam.cpp [hunk 6]
```diff
@@ -12,57 +12,171 @@
...
+        } else {
+          state->max_exp_avg_sq({});
```
Comment (yf225) [inline torch/csrc/api/src/optim/adam.cpp]: Since we are setting default value of `max_exp_avg_sq` to undefined tensor, we probably don't need to explicitly set its value to undefined tensor here.

File: torch/csrc/api/src/optim/adam.cpp [hunk 7]
```diff
@@ -12,57 +12,171 @@
...
+        }
+        state_[c10::guts::to_string(p.unsafeGetTensorImpl())] = std::move(state);
+      }
+      auto state = static_cast<AdamParamState&>(*state_[c10::guts::to_string(p.unsafeGetTensorImpl())]);
+      // Exponential moving average of gradient values
```
Comment (yf225) [inline torch/csrc/api/src/optim/adam.cpp]: ah I see that we moved the comments here, but maybe we should keep them the same place as Python code :D

File: torch/csrc/api/src/optim/adam.cpp [hunk 8]
```diff
@@ -12,57 +12,171 @@
...
+      // Decay the first and second moment running average coefficient
+      exp_avg.mul_(options.beta1()).add_(grad, 1 - options.beta1());
+      exp_avg_sq.mul_(options.beta2()).addcmul_(grad, grad, 1 - options.beta2());
 
-    const auto step_size =
-        options.learning_rate() / bias_correction1;
+      Tensor denom;
+      if(options.amsgrad()) {
+        // Maintains the maximum of all 2nd moment running avg. till now
+        at::max_out(max_exp_avg_sq, exp_avg_sq, max_exp_avg_sq);
```
Comment (yf225) [inline torch/csrc/api/src/optim/adam.cpp]: maybe `torch::max_out`, since we are in torch namespace now :D

File: torch/csrc/api/src/optim/adam.cpp [hunk 9]
```diff
@@ -12,57 +12,171 @@
...
+  param_groups_.emplace_back(OptimizerParamGroup(parameters, defaults_->clone()));
+}
+
+const std::vector<Tensor>& Adam::parameters() const noexcept {
+  return param_groups_.at(0).params();
+}
+
+std::vector<Tensor>& Adam::parameters() noexcept {
+  return param_groups_.at(0).params();
+}
```
Comment (yf225) [inline torch/csrc/api/src/optim/adam.cpp]: The content of the above three functions `add_parameters`  / const ref `parameters` / non-const ref `parameters` seem to always be the same for all optimizers, and maybe we should create functions similar to `_size_new_design` and have all optimizers call the common functions instead :D

File: torch/csrc/api/src/optim/adam.cpp [hunk 10]
```diff
@@ -12,57 +12,171 @@
...
+    torch::optim::serialize(archive, "exp_average_buffers", exp_average_buffers);
+    torch::optim::serialize(archive, "exp_average_sq_buffers", exp_average_sq_buffers);
+    torch::optim::serialize(archive, "max_exp_average_sq_buffers", max_exp_average_sq_buffers);
+    // since there were no param_groups prior to version 1.5.0, assuming all tensors are now in one param_group
+    std::vector<Tensor> params = param_groups_.at(0).params();
+    for (size_t idx = 0; idx < step_buffers.size(); idx++) {
+      auto state = std::make_unique<AdamParamState>();
+      state->step(step_buffers[idx]);
+      state->exp_avg(exp_average_buffers[idx]);
+      state->exp_avg_sq(exp_average_sq_buffers[idx]);
```
Comment (yf225) [inline torch/csrc/api/src/optim/adam.cpp]: To index a vector in this function, I'd recommend using `step_buffers.at(idx)` instead of `step_buffers[idx]`. The reason is that `step_buffers.at(idx)` checks whether `idx` is out of bound for `step_buffers`, and gives a meaningful error message when the out-of-bound error happens (versus `step_buffers[idx]` which returns garbage data if `idx` is out of bound)

File: test/cpp/api/serialize.cpp [hunk 11]
```diff
@@ -553,6 +552,56 @@ TEST(SerializeTest, Optim_SGD) {
...
+  for (size_t i = 0; i < params_.size(); i++) {
+    if(i != (params_.size() - 1)) {
+      auto key_ = c10::guts::to_string(params_[i].unsafeGetTensorImpl());
+      const AdamParamState& curr_state_ = static_cast<const AdamParamState&>(*(optim1_state.at(key_).get()));
+      step_buffers.emplace_back(curr_state_.step());
+      exp_average_buffers.emplace_back(curr_state_.exp_avg());
+      exp_average_sq_buffers.emplace_back(curr_state_.exp_avg_sq());
+      if(curr_state_.max_exp_avg_sq().defined()) {
+        max_exp_average_sq_buffers.emplace_back(curr_state_.max_exp_avg_sq());
+      }
```
[the remark this answers is not in the document]
Comment (anjali411) [reply test/cpp/api/serialize.cpp]: yeah but the old serialization logic did that

File: torch/csrc/api/src/optim/adam.cpp [hunk 12]
```diff
@@ -12,57 +12,167 @@
...
-    if (options.amsgrad()) {
-      auto& max_exp_average_sq = buffer_at(max_exp_average_sq_buffers, i);
-      max_exp_average_sq = torch::max(max_exp_average_sq, exp_average_sq);
-      denom = max_exp_average_sq / bias_correction2;
-    } else {
-      denom = exp_average_sq / bias_correction2;
-    }
+      state.step(state.step()+1);
+      auto beta1 = std::get<0>(options.betas());
+      auto beta2 = std::get<0>(options.betas());
```
Comment (yf225) [inline torch/csrc/api/src/optim/adam.cpp]: I think this should be
```suggestion
      auto beta2 = std::get<1>(options.betas());
```

File: torch/csrc/api/include/torch/optim/adam.h [hunk 13]
```diff
@@ -18,44 +18,62 @@ class InputArchive;
...
-  explicit Adam(ParameterContainer&& parameters, const AdamOptions& options_)
-      : Optimizer(std::forward<ParameterContainer>(parameters)),
-        options(options_) {}
+   explicit Adam(std::vector<OptimizerParamGroup> param_groups,
+       AdamOptions defaults) : Optimizer(std::move(param_groups), std::make_unique<AdamOptions>(defaults)) {
+     TORCH_CHECK(defaults.lr() >= 0, "Invalid learning rate: ", defaults.lr());
+     TORCH_CHECK(defaults.eps() >= 0, "Invalid epsilon value: ", defaults.eps());
+     auto betas = defaults.betas();
+     TORCH_CHECK(std::get<0>(betas) >= 0, "Invalid learning rate: ", std::get<0>(betas));
+     TORCH_CHECK(std::get<1>(betas) >= 0, "Invalid learning rate: ", std::get<0>(betas));
```
Comment (yf225) [inline torch/csrc/api/include/torch/optim/adam.h]:
```suggestion
     TORCH_CHECK(std::get<1>(betas) >= 0, "Invalid learning rate: ", std::get<1>(betas));
```

File: test/cpp/api/serialize.cpp [hunk 14]
```diff
@@ -553,6 +558,56 @@ TEST(SerializeTest, Optim_SGD) {
   is_optimizer_state_equal<SGDParamState>(optim1.state(), optim1_2.state());
 }
 
+TEST(SerializeTest, Optim_Adam) {
+  test_serialize_optimizer<Adam, AdamOptions, AdamParamState>(AdamOptions().amsgrad(true).weight_decay(0.5).betas(std::make_tuple(0.6, 0.999)));
```
Comment (anjali411) [inline test/cpp/api/serialize.cpp]: there's a test failure at this line: https://github.com/pytorch/pytorch/blob/1546d2afeb98fcd7bc5b58261d6e31ad7794e833/test/cpp/api/serialize.cpp#L154 any idea why @vincentqb 
p.s. the step value is being incremented correctly

Comment (anjali411) [reply test/cpp/api/serialize.cpp]: P.S. we have a similar check for optimizers in Python

Comment (vincentqb) [reply test/cpp/api/serialize.cpp]: For reference, from [here](https://app.circleci.com/jobs/github/pytorch/pytorch/4697533):
```
[ RUN      ] SerializeTest.Optim_Adam
step1 model1
step2 model1
step1 model2
step1 model2
step1 model3
..\test\cpp\api\serialize.cpp(159): error: Value of: param1[name].norm().item<float>() != param2[name].norm().item<float>()
  Actual: false
Expected: true
[  FAILED  ] SerializeTest.Optim_Adam (6 ms)
```

Can you link to the equivalent in python?

Comment (vincentqb) [reply test/cpp/api/serialize.cpp]: Do you know which commit in this PR caused this?

Comment (anjali411) [reply test/cpp/api/serialize.cpp]: resolved: optimizer wasn't taking big enough step and the value exp_avg/denom * - step_size,  was coming out to be very close for model1 step2 and model2 step1 (with optim2_2).  step_size = options.lr() / bias_correction1 and bias_correction1 is a function of step.val(), hence increasing the lr increased the step_size in both cases but by different amounts and hence increased the difference in values in the two cases.

File: torch/csrc/api/include/torch/optim/serialize.h [hunk 15]
```diff
@@ -214,13 +214,21 @@ void serialize(
...
 #define _TORCH_OPTIM_SERIALIZE_TORCH_ARG(name) \
-  archive.write(#name, IValue(name()))
+  archive.write(#name, IValue(name())); \
+
+// serialize only if the tensor is defined
+#define _TORCH_OPTIM_SERIALIZE_TORCH_ARG_IF_DEFINED(name) \
+{ \
+  if(name().defined()) \
+    _TORCH_OPTIM_SERIALIZE_TORCH_ARG(name); \
+} \
```
Comment (yf225) [inline torch/csrc/api/include/torch/optim/serialize.h]: It's a little sad that we need two macros `_TORCH_OPTIM_SERIALIZE_TORCH_ARG_IF_DEFINED` and `_TORCH_OPTIM_SERIALIZE_TORCH_ARG` to support serializing Tensor and non-Tensor objects. But actually we can merge them into one:
```cpp
#define _TORCH_OPTIM_SERIALIZE_TORCH_ARG(name) \
{ \
  auto ivalue = torch::IValue(name()); \
  if (!(ivalue.isTensor() && ivalue.unsafeToTensorImpl() == at::UndefinedTensorImpl::singleton())) { \
    archive.write(#name, ivalue); \
  } \
}
```

File: torch/csrc/api/src/optim/adam.cpp [hunk 16]
```diff
@@ -12,57 +12,173 @@
...
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, eps);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, weight_decay);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(bool, amsgrad);
+}
+
+bool operator==(const AdamParamState& lhs, const AdamParamState& rhs) {
+  bool is_max_exp_avg_sq_defined_and_equal = !(lhs.max_exp_avg_sq().defined() ^ rhs.max_exp_avg_sq().defined());
+  if(!is_max_exp_avg_sq_defined_and_equal) {
+    is_max_exp_avg_sq_defined_and_equal = torch::equal(lhs.max_exp_avg_sq(), rhs.max_exp_avg_sq());
+  }
```
Comment (yf225) [inline torch/csrc/api/src/optim/adam.cpp]: When `lhs.max_exp_avg_sq()` is undefined and `rhs.max_exp_avg_sq()` is defined, `!(lhs.max_exp_avg_sq().defined() ^ rhs.max_exp_avg_sq().defined())` gives `false`, and `! is_max_exp_avg_sq_defined_and_equal` is `true`, which means we would go into the if branch and run `torch::equal` on undefined tensors (!). It might be easier to read and check for correctness if we write it as:
```cpp
  return (lhs.step() == rhs.step()) &&
          torch::equal(lhs.exp_avg(), rhs.exp_avg()) &&
          torch::equal(lhs.exp_avg_sq(), rhs.exp_avg_sq()) &&
          ((!lhs.max_exp_avg_sq().defined() && !rhs.max_exp_avg_sq().defined()) ||
           (lhs.max_exp_avg_sq().defined() && rhs.max_exp_avg_sq().defined() && torch::equal(lhs.max_exp_avg_sq(), rhs.max_exp_avg_sq())));
```

**Review by yf225 — APPROVED** [review]
Thanks a lot for the awesome work!
